rs/index/src/ivf.rs

```diff
@@ -1,6 +1,145 @@
+        let flattened_dataset: Vec<f32> = dataset.iter().flatten().cloned().collect();
+        let dim = dataset[0].len();
+        let kmeans: KMeans<_, 8> = KMeans::new(flattened_dataset, dataset.len(), dim);
+        // TODO(tyb0807): maybe have IvfConfig
+        let result = kmeans.kmeans_minibatch(
+            4, // batch_size
+            num_clusters,
+            1000, // max_iter
+            KMeans::init_random_sample,
+            &config,
+        );
+        debug!("Error: {}", result.distsum);
+        let centroids = result
+            .centroids
+            .chunks(dim)
+            .map(|chunk| chunk.to_vec())
+            .collect();
+
+        let mut inverted_lists = HashMap::new();
+        // Assign vectors to nearest centroids
+        for (i, vector) in dataset.iter().enumerate() {
+            let nearest_centroid = Self::find_nearest_centroid(&vector, &centroids);
+            inverted_lists
+                .entry(nearest_centroid)
+                .or_insert_with(Vec::new)
+                .push(i);
+        }
+        Self {
+            dataset,
+            num_clusters,
+            centroids,
+            inverted_lists,
+            num_probes,
+        }
+    }
+
+    fn find_nearest_centroid(vector: &Vec<f32>, centroids: &Vec<Vec<f32>>) -> usize {
+        let mut calculator = L2DistanceCalculator::new();
+        centroids
+            .iter()
+            .enumerate()
+            .min_by(|(_, a), (_, b)| {
+                let dist_a = calculator.calculate(&vector, &a);
+                let dist_b = calculator.calculate(&vector, &b);
+                dist_a.partial_cmp(&dist_b).unwrap()
+            })
+            .map(|(index, _)| index)
+            .unwrap()
+    }
+
+    fn search(&self, query: &[f32], k: usize) -> Vec<(usize, f32)> {
+        let mut results = Vec::new();
+        let mut probed_centroids = Vec::new();
+
+        // Find the nearest centroids to the query.
+        for _ in 0..self.num_probes {
+            let nearest_centroid = Self::find_nearest_centroid(&query.to_vec(), &self.centroids);
+            if !probed_centroids.contains(&nearest_centroid) {
+                probed_centroids.push(nearest_centroid);
+            }
+        }
+
+        // Search in the inverted lists of the nearest centroids.
+        for &centroid in &probed_centroids {
+            if let Some(list) = self.inverted_lists.get(&centroid) {
+                for &idx in list {
+                    let distance = L2DistanceCalculator::new().calculate(query, &self.dataset[idx]);
+                    results.push((idx, distance));
+                }
+            }
+        }
+
+        // Sort results and return top k.
+        results.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
+        results.truncate(k);
+        results
+    }
+}
+
+//impl Index for Ivf {
+//    fn search(&self, query: &[f32], k: usize) -> Option<Vec<u64>> {
+//        Some(self.ann_search(query, k, 10))
+//    }
+//}
```

Comment:
let have this impl for `Index` as well

Reply:
Same, this means we need to have mmap support in IVF and not storing the whole dataset (IIUC?)

Reply:
I don't think so. It's just a trait